Gateway unit test flake

Occasional failures in unit test - github.com/hyperledger/fabric/internal/pkg/gateway

```
--- FAIL: TestEndorse (0.28s)
    --- FAIL: TestEndorse/context_timeout_collecting_endorsements (0.01s)
        api_test.go:2221: 
            	Error Trace:	/home/vsts/work/1/fabric/internal/pkg/gateway/api_test.go:2221
            	            				/home/vsts/work/1/fabric/internal/pkg/gateway/api_test.go:1041
            	Error:      	Error "rpc error: code = Aborted desc = failed to collect enough transaction endorsements, see attached details for more info" does not contain "endorsement timeout expired while collecting endorsements"
            	Test:       	TestEndorse/context_timeout_collecting_endorsements
            	Messages:   	error string
```

An example at:

https://dev.azure.com/Hyperledger/Fabric/_build/results?buildId=56894&view=logs&jobId=6b58850f-3858-5a05-33e2-5e41cbf03c4e&j=6b58850f-3858-5a05-33e2-5e41cbf03c4e&t=bddec1cf-ba37-5883-9c3e-fd1e8608f9a1

--- a/internal/pkg/gateway/api_test.go
+++ b/internal/pkg/gateway/api_test.go
@@ -153,6 +153,7 @@ type testDef struct {
 type preparedTest struct {
 	server         *Server
 	ctx            context.Context
+	cancel         context.CancelFunc
 	signedProposal *peer.SignedProposal
 	localEndorser  *mocks.EndorserClient
 	discovery      *mocks.Discovery
@@ -477,14 +478,15 @@ func TestEvaluate(t *testing.T) {
 				"g1": {{endorser: localhostMock, height: 3}}, // msp1
 			},
 			postSetup: func(t *testing.T, def *preparedTest) {
-				var cancel context.CancelFunc
-				def.ctx, cancel = context.WithTimeout(def.ctx, 100*time.Millisecond)
+				def.ctx, def.cancel = context.WithTimeout(def.ctx, 100*time.Millisecond)
 
 				def.localEndorser.ProcessProposalStub = func(ctx context.Context, proposal *peer.SignedProposal, option ...grpc.CallOption) (*peer.ProposalResponse, error) {
-					cancel()
 					time.Sleep(200 * time.Millisecond)
 					return createProposalResponse(t, peer1Mock.address, "mock_response", 200, ""), nil
 				}
+			},
+			postTest: func(t *testing.T, def *preparedTest) {
+				def.cancel()
 			},
 			errCode:   codes.DeadlineExceeded,
 			errString: "evaluate timeout expired",
@@ -1000,14 +1002,15 @@ func TestEndorse(t *testing.T) {
 				"g2": {{endorser: peer4Mock, height: 5}},     // msp3
 			},
 			postSetup: func(t *testing.T, def *preparedTest) {
-				var cancel context.CancelFunc
-				def.ctx, cancel = context.WithTimeout(def.ctx, 100*time.Millisecond)
+				def.ctx, def.cancel = context.WithTimeout(def.ctx, 100*time.Millisecond)
 
 				def.localEndorser.ProcessProposalStub = func(ctx context.Context, proposal *peer.SignedProposal, option ...grpc.CallOption) (*peer.ProposalResponse, error) {
-					cancel()
 					time.Sleep(200 * time.Millisecond)
 					return createProposalResponse(t, peer1Mock.address, "mock_response", 200, ""), nil
 				}
+			},
+			postTest: func(t *testing.T, def *preparedTest) {
+				def.cancel()
 			},
 			errCode:   codes.DeadlineExceeded,
 			errString: "endorsement timeout expired while collecting first endorsement",
@@ -1019,14 +1022,15 @@ func TestEndorse(t *testing.T) {
 				"g2": {{endorser: peer4Mock, height: 5}},     // msp3
 			},
 			postSetup: func(t *testing.T, def *preparedTest) {
-				var cancel context.CancelFunc
-				def.ctx, cancel = context.WithTimeout(def.ctx, 100*time.Millisecond)
+				def.ctx, def.cancel = context.WithTimeout(def.ctx, 100*time.Millisecond)
 
 				peer4Mock.client.(*mocks.EndorserClient).ProcessProposalStub = func(ctx context.Context, proposal *peer.SignedProposal, option ...grpc.CallOption) (*peer.ProposalResponse, error) {
-					cancel()
 					time.Sleep(200 * time.Millisecond)
 					return createProposalResponse(t, peer4Mock.address, "mock_response", 200, ""), nil
 				}
+			},
+			postTest: func(t *testing.T, def *preparedTest) {
+				def.cancel()
 			},
 			errCode:   codes.DeadlineExceeded,
 			errString: "endorsement timeout expired while collecting endorsements",

